Addition of an Extended Holdings layout for items on the holdings record tab

jlcooper opened the pull request:
As a side effect of discussions around VUFIND issue 1269, I thought I'd have a go at adding in an option to use an extended layout for items in the holdings record tab.  As it's now at the first draft stage I thought I'd submit a pull request for comment.

If the `extendedHoldingFields` is set to true in `config.ini` then the new `extendedholdingsils.phtml` will be used for each item row. If not set or set to false then the existing layout will be used.

I've used flexboxes for the CSS of the individual fields so that it should be easy to disable/enable, size and reorder them via CSS.

demiankatz commented:
Thanks for sharing this -- definitely a good starting point for conversation. I haven't had a time to take a close look yet, but a couple of quick comments:

1.) Rather than putting the extendedHoldingFields configuration check into VuFind\Controller\AbstractRecord, I would instead make it part of VuFind\RecordTab\HoldingsILS (which I realize will also require a change to VuFind\RecordTab\Factory::getHoldingsILS to inject the setting into the tab object). Since this setting only applies to one tab, I think it's better to set it up in a more specific location to keep the deeper infrastructure code simpler.

2.) I'm interesting in @crhallberg's feedback on the overall style approach, since he's been the lead on the VuFind UI design for a while and might have some useful input.

I'll try to actually test this out in the near future and will let you know if I have more feedback at that point... but I wanted to get the ball rolling as soon as I could. Thanks again for the work on this!

jlcooper commented:
I've moved the extendedHoldingFields configuration check into VuFind\RecordTab\HoldingsILS and VuFind\RecordTab\Factory::getHoldingsILS.

crhallberg commented on themes/bootprint3/less/bootprint.less:
```diff
@@ -167,3 +167,34 @@ ul.random.image li img { margin: 0 auto; }
 .jstree-node.active { background-color: @active-orange; }
+
+/* --- extended holdings --- */
+
```
These styles seem identical to the ones in bootstrap.less. Bootprint will inherit from Bootstrap, so you don't need it in both places.

crhallberg commented on themes/sandal/less/sandal.less:
```diff
@@ -232,3 +232,34 @@ header {
 }
+
+/* --- extended holdings --- */
+
```
Same thing here: these styles are in Bootstrap and will be inherited.

crhallberg commented on themes/bootstrap3/templates/RecordTab/extendedholdingsils.phtml:
```diff
@@ -0,0 +1,84 @@
+          <? /* Declare that the item is to be borrowed, not for sale */ ?>
+            <link property="businessFunction" href="http://purl.org/goodrelations/v1#LeaseOut" />
+            <link property="itemOffered" href="#record" />
+            <!-- <? var_dump($this->holding) ?> -->
```
This is a nice debug. Just have to remember to remove it before merge.

crhallberg commented on themes/bootstrap3/templates/RecordTab/extendedholdingsils.phtml:
```diff
@@ -0,0 +1,84 @@
+            </span>
+            <span class="holding-field place-hold">
+              <? if (isset($this->holding['link']) && $this->holding['link']): ?>
+                <a class="<?=$check ? 'checkRequest ' : ''?>placehold" <? if (!empty($this->holding['linkLightbox'])): ?>data-lightbox <? endif; ?>href="<?=$this->recordLink()->getRequestUrl($this->holding['link'])?>"><i class="fa fa-flag" aria-hidden="true"></i>&nbsp;<?=$this->transEsc($check ? "Check Hold" : "Place a Hold")?></a>
```
These check variables don't seem to be initialized. You can [copy the setup from holdingsils](https://github.com/vufind-org/vufind/blob/master/themes/bootstrap3/templates/RecordTab/holdingsils.phtml#L92) replacing `$row` with `$this->holding`.

crhallberg replied:
```php
<?
  // AJAX Check record?
  $check = isset($this->holding['check']) && $this->holding['check'];
  $checkStorageRetrievalRequest = isset($this->holding['checkStorageRetrievalRequest']) && $this->holding['checkStorageRetrievalRequest'];
  $checkILLRequest = isset($this->holding['checkILLRequest']) && $this->holding['checkILLRequest'];
?>
```

crhallberg replied:
Perfect.

crhallberg commented on themes/bootstrap3/templates/RecordTab/holdingsils.phtml:
```diff
@@ -96,67 +98,73 @@
-              <? endif; ?>
+        <? if ($this->tab->extendedHoldingFields):?>
+
+          <?=$this->context($this)->renderInContext('RecordTab/extendedholdingsils.phtml', [holding => $row]); ?>
```
I got an error on this line with `holding`. You most likely meant `"holding"`.

demiankatz commented on themes/bootstrap3/templates/RecordTab/holdingsils.phtml:
```diff
@@ -97,67 +97,73 @@
-              <? if (isset($row['ILLRequestLink']) && $row['ILLRequestLink']): ?>
-                <a class="<?=$checkILLRequest ? 'checkILLRequest ' : ''?>placeILLRequest" data-lightbox href="<?=$this->recordLink()->getRequestUrl($row['ILLRequestLink'])?>"><i class="fa fa-flag" aria-hidden="true"></i>&nbsp;<?=$this->transEsc($checkILLRequest ? "ill_request_check_text" : "ill_request_place_text")?></a>
-              <? endif; ?>
+        <? if ($this->tab->extendedHoldingFields):?>
```
Here's a thought to make things a little more flexible: what if, instead of having a boolean extendedHoldingsFields configuration, you had a string holdingsTemplate configuration. This could be "standard" or "extended", for example. The code could then load RecordTab/holdingsils/standard.phtml or RecordTab/holdingsils/extended.phtml or whatever. There could be a try..catch around the loading to force it to fail back to standard if an invalid template is specified. Obviously this would also require all the logic currently in the else clause to be refactored to a separate template... but I think in the end that might make everything a little more readable and extensible. What do you think?

jlcooper replied:
It would make it easier for others to add their own custom holdings structure if they want.  I'll try to find some time to have a look at implementing it.

demiankatz commented:
Thanks, @jlcooper, this looks great! I'll try to find time to do some testing next week (right now we're still dealing with weather-related issues over here). I'm also interested to hear whether @crhallberg has any further thoughts on this.

demiankatz commented:
@jlcooper, I took a closer look at this today. (I also merged the latest master into the branch to resolve some conflicts caused by this week's changes). The refactoring of the standard view seems to work well for me. I noticed a couple of issues with the extended view:

- When I logged in and accessed an item that was allowed to be placed on hold, the hold link displayed twice.
- The columns don't always align in the expected way; I believe @crhallberg may have noticed this as well.

Also, just a general suggestion: now that standard.phtml and extended.phtml are their own templates, you can probably strip the leading indentation off of the files to improve readability. (If you don't have a convenient tool to do this yourself, I'm happy to help... but usually shift-tab will do the trick in most editors).

crhallberg commented:
I did notice the alignment issue. I think the issue with the columns is that you're introducing a flex column system to an existing table. I don't think the columns will ever line up in this case.

Is there a reason you went with flex over working with the existing table?

jlcooper commented:
I tried to take another look at this today, but after pulling down the latest branch version with the merged changes from master I get the following error:

```
CRIT (2): VuFindSearch\Backend\Exception\RemoteErrorException : 500 {metadata={error-class=org.apache.solr.common.SolrException,root-error-class=org.apache.solr.common.SolrException},msg=SolrCore 'biblio' is not available due to init failure: Could not load conf for core biblio: Can't load schema /vagrant/solr/vufind/biblio/conf/schema.xml: Field type date{class=org.apache.solr.schema.DatePointField,analyzer=org.apache.solr.schema.FieldType$DefaultAnalyzer,args={omitNorms=true, sortMissingLast=true, class=solr.DatePointField}} does not support doc values,trace=org.apache.solr.common.SolrException: SolrCore 'biblio' is not available due to init failure: Could not load conf for core biblio: Can't load schema /vagrant/solr/vufind/biblio/conf/schema.xml: Field type date{class=org.apache.solr.schema.DatePointField,analyzer=org.apache.solr.schema.FieldType$DefaultAnalyzer,args={omitNorms=true, sortMissingLast=true, class=solr.DatePointField}} does not support doc values?   at org.apache.solr.core.CoreContainer.getCore(CoreContainer.java:1151)? at org.apache.solr.servlet.HttpSolrCall.
```

I've run the usual `composer install` to make sure I've got the latest dependencies (including solr 7.2.1), but still get the issue.  I also get the issue on the latest branch.  I haven't tried destroying my vagrant machine and re-provisioning it yet in case it's an issue with upgrading to the latest solr version that we need to be address.

Once we've got this figured out I'll be able to get back to fixing the duplicate hold link and the CSS for the flexboxes.

demiankatz commented:
@jlcooper, I just pulled this branch on my test server, merged the latest master into it, spun it up, and everything seems fine. However, I don't usually test using Vagrant, so maybe that is a factor. Do you mind destroying your vagrant machine and trying again? If that doesn't fix it, let me know and I'll see if I can reproduce the Vagrant problem on my end... but I suspect if there is a problem here, it's Vagrant-related rather than more generalized.

jlcooper commented:
I destroyed the vagrant machine and re-provisioned it, but it didn't make any difference.  In the end I had to remove the contents of the `solr/vendor` directory and rerun `composer install` to reinstall solr.  Afterwards it worked fine (I suspect that the previous `composer install` had encountered a locked file or something when trying to update it and left it in an invalid status).

I think I've fixed the issue with `Recall this` and `Place a hold`  both appearing for the same item and I've also added in some styling for the flexboxes to improve their alignment.

@crhallberg, there were two main reason's that I went with flex boxes. The first reason was more tactical, in that we had more columns of item information to display than we had columns in the table to play with.  The second reason was more strategical, in that it allows users to order their columns in their CSS rather than having to have their own copy of the template.

demiankatz commented:
Thanks, @jlcooper, the duplicate hold link problem does seem to be resolved on my end. However, I'm now seeing an issue where text overflows the tab when all request options are enabled (most easily seen by using the Demo ILS driver and turning on all the settings in the [Catalog] section of config.ini). Screen shot:

![image](https://user-images.githubusercontent.com/309069/38431986-0b6ae588-3994-11e8-9e1d-ea98322bd127.png)

Perhaps this is an unavoidable extreme case, but it would be nice if things could wrap better.

demiankatz commented:
@jlcooper, just letting you know I haven't forgotten about this one. I made a few adjustments (including merging in the latest master), but I still have a few more edits I'd like to make to further simplify the template code -- hopefully I'll have time in the near future to continue work.

demiankatz commented:
@jlcooper, I made a few more adjustments today -- rather than calculating the $check variables in the top-level tab template, I moved that calculation into the standard/extended templates. Even though that duplicates some logic, I thought it was cleaner to pass just one value into the sub-templates, and it saves us from being locked in to specific names for those variables. I also removed the $this from all of the references to $holding (since it is not strictly necessary and improves readability), and I used null coalescing to simplify many of the if conditions. I think this gives us a much cleaner and more flexible holdings display template!

Do you mind taking one more look at this and letting me know if you see anything amiss? Once you're happy, I think this is probably ready to be merged (though I'll do a little more testing on this end too, just to be sure).

Thanks again for moving this forward!

demiankatz commented:
One other question for @crhallberg and @jlcooper: do you like leaving the `<tr>` outside the holdings templates, or do you think it would be cleaner to move it into the sub-templates so that they are rendering a more complete unit? I could go either way, but it seems worth considering the possibilities now before we commit.

jlcooper commented:
@demiankatz, I can see a case for moving the `<tr>`  into the sub-templates as that would enable sub-templates to choose to not add anything for an entry, it would also make it a bit clearer if a sub-template wanted to add multiple rows per item.

@crhallberg, does that make sense to you? If so then I'll look at moving them over.

crhallberg commented:
While I'm cautious about putting other elements into a table, I agree with the argument that a template that would want to add multiple rows should have the opening tr. I'm in favor of moving it in.

jlcooper commented:
I've moved the `<tr>` tag into the sub-templates.

@demiankatz, @crhallberg - while doing that I also wondered about the check that the barcode exists for an item before displaying it. Do you think this would be something that makes more sense to be in the sub-templates as well?

demiankatz commented:
@jlcooper, good point, I think moving the barcode check inside does make sense as well.

demiankatz commented:
Thanks again, @jlcooper -- merged following a few more minor last-minute tweaks.
